src/backend/distributed/utils/background_jobs.c

```diff
@@ -0,0 +1,885 @@
+
+#include "postgres.h"
+
+#include "safe_mem_lib.h"
+
+#include "libpq-fe.h"
+#include "libpq/pqformat.h"
+#include "libpq/pqmq.h"
+#include "libpq/pqsignal.h"
+#include "parser/analyze.h"
+#include "pgstat.h"
+#include "storage/dsm.h"
+#include "storage/ipc.h"
+#include "storage/shm_mq.h"
+#include "storage/shm_toc.h"
+#include "tcop/pquery.h"
+#include "tcop/tcopprot.h"
+#include "tcop/utility.h"
+#include "utils/backend_status.h"
+#include "utils/memutils.h"
+#include "utils/portal.h"
+#include "utils/ps_status.h"
+#include "utils/resowner.h"
+#include "utils/snapmgr.h"
+#include "utils/timeout.h"
+
+#include "distributed/background_jobs.h"
+#include "distributed/citus_safe_lib.h"
+#include "distributed/listutils.h"
+#include "distributed/maintenanced.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/metadata_utility.h"
+#include "distributed/shard_cleaner.h"
+#include "distributed/resource_lock.h"
+
+bool BackgroundTaskMonitorDebugDelay = false;
+
+/* Table-of-contents constants for our dynamic shared memory segment. */
+#define CITUS_BACKGROUND_TASK_MAGIC 0x51028081
+#define CITUS_BACKGROUND_TASK_KEY_DATABASE 0
+#define CITUS_BACKGROUND_TASK_KEY_USERNAME 1
+#define CITUS_BACKGROUND_TASK_KEY_COMMAND 2
+#define CITUS_BACKGROUND_TASK_KEY_QUEUE 3
+#define CITUS_BACKGROUND_TASK_NKEYS 4
+
+static BackgroundWorkerHandle * StartCitusBackgroundJobExecuter(char *database,
+																char *user, char *command,
+																dsm_segment **pSegment);
+static void ExecuteSqlString(const char *sql);
+static void ConsumeTaskWorkerOutput(shm_mq_handle *responseq, BackgroundTask *task,
+									bool *hadError);
+
+static void
+BackgroundTaskUpdatePid(BackgroundTask *task, pid_t *pid)
+{
+	if (pid)
+	{
+		if (!task->pid)
+		{
+			MemoryContext taskContext = GetMemoryChunkContext(task);
+			task->pid = MemoryContextAlloc(taskContext, sizeof(int32));
+		}
+		*task->pid = *pid;
```

Comment:
yeah, am not supper happy with that either, there are a couple, `pid`/`started_at`/`finished_at`/`not_before`/ etc. I will check if 0 would be good enough for all.

Otherwise I will probably add a struct at the end with fixed places we can point to that pfree all at once with the allocated struct. But still will have a pointer that can be `NULL` separate from the value. Although I think that `0` would probably be good enough

Reply:
for some the value 0 is hard to distingiush from their `NULL` value, I added a struct to the end of both `BackgrounJob` and `BackroundTask` to hold the values for things like count's and timestamp's. This makes that we don't need the deepfree-ing anymore either.

To make the (un)setting easy I have added two macro's that populate the marked structs at the end.